service/src/main/java/uk/nhs/adaptors/gpc/consumer/filters/SdsFilter.java

```diff
@@ -0,0 +1,162 @@
+        if (StringUtils.isBlank(System.getProperty(GPC_URL_ENVIRONMENT_VARIABLE))) {
+            ServerHttpRequest serverHttpRequest = exchange.getRequest();
+            extractInteractionId(serverHttpRequest.getHeaders())
+                .ifPresent(id -> proceedSdsLookup(serverHttpRequest, exchange, id));
+        }
+
+        return chain.filter(exchange);
+    }
+
+    @Override
+    public int getOrder() {
+        return RouteToRequestUrlFilter.ROUTE_TO_URL_FILTER_ORDER + 1;
+    }
+
+    @PostConstruct
+    @SuppressWarnings("unused")
+    public void initializeSdsRequestFunctions() {
+        sdsRequestFunctions = Map.of(
+            STRUCTURED_ID, sdsClient::callForGetStructuredRecord,
+            PATIENT_SEARCH_ID, sdsClient::callForPatientSearchAccessDocument,
+            DOCUMENT_SEARCH_ID, sdsClient::callForSearchForDocumentRecord,
+            BINARY_READ_ID, sdsClient::callForRetrieveDocumentRecord);
+    }
+
+    private void proceedSdsLookup(ServerHttpRequest serverHttpRequest,
+            ServerWebExchange exchange,
+            String integrationId) {
+        Optional<SdsClient.SdsResponseData> response
+            = performRequestAccordingToInteractionId(integrationId, serverHttpRequest);
+
+        if (response.isPresent()) {
+            String address = response.get()
+                .getAddress();
+            prepareLookupUri(address, serverHttpRequest.getPath()).ifPresent(uri -> exchange.getAttributes()
+                .put(ServerWebExchangeUtils.GATEWAY_REQUEST_URL_ATTR, uri));
+        } else {
+            LOGGER.error("SDS filter request was unsuccessful.");
+        }
+    }
+
+    private Optional<SdsClient.SdsResponseData> performRequestAccordingToInteractionId(String interactionId,
+            ServerHttpRequest serverHttpRequest) {
+        if (isRequestFunctionAvailable(interactionId)) {
+            Optional<String> organisation = extractOrganisation(serverHttpRequest
+                .getPath());
+
+            if (organisation.isPresent()) {
+                LOGGER.info("Performing request with organisation \"{}\" and NHS service endpoint id \"{}\"",
+                    organisation.get(), interactionId);
+                return sdsRequestFunctions.get(interactionId)
+                    .apply(organisation.get());
+            }
+        }
+        return Optional.empty();
+    }
+
+    private boolean isRequestFunctionAvailable(String interactionId) {
+        return sdsRequestFunctions.keySet()
+            .stream()
+            .anyMatch(key -> key.equals(interactionId));
+    }
+
+    private Optional<String> extractOrganisation(RequestPath requestPath) {
+        String path = requestPath.toString();
+        if (path.contains(SLASH)) {
+            String[] pathElements = path.split(SLASH);
+
+            if (pathElements.length > 1) {
+                return Optional.of(pathElements[1]);
+            }
+        }
+        return Optional.empty();
```

Comment:
Using "elements()" may be more robust than manually splitting on slash.

```suggestion
                return requestPath.elements().stream().skip(1).findFirst().orElseThrow("URL does not contain ODS code in its second element").map(PathContainer.Element::value);
```